MetaMorpheus/EngineLayer/FdrAnalysis/FdrAnalysisEngine.cs

```diff
@@ -152,54 +155,82 @@ private void DoFalseDiscoveryRateAnalysis(FdrAnalysisResults myAnalysisResults)
-            double[] cumulativeTargetPerNotch = new double[MassDiffAcceptorNumNotches + 1];
-            double[] cumulativeDecoyPerNotch = new double[MassDiffAcceptorNumNotches + 1];
+            //set up arrays for local FDRs
+            var allNotches = psms
+                .SelectMany(p => p.BestMatchingBioPolymersWithSetMods.Select(h => h.Notch))
+                .ToHashSet();
+
+            var cumulativeTargetPerNotch = allNotches.ToDictionary(n => n, _ => 0.0);
+            var cumulativeDecoyPerNotch = allNotches.ToDictionary(n => n, _ => 0.0);
 
             //Assign FDR values to PSMs
+            List<IGrouping<int, SpectralMatchHypothesis>> ambigResultsGroupedByNotch = new(8);
             foreach (var psm in psms)
             {
-                // Stop if canceled
                 if (GlobalVariables.StopLoops) { break; }
 
-                // we have to keep track of q-values separately for each notch
-                int notch = psm.Notch ?? MassDiffAcceptorNumNotches;
-                if (psm.IsDecoy)
-                {
-                    // in that case, count it as the fraction of decoy hits
-                    // e.g. if the PSM matched to 1 target and 2 decoys, it counts as 2/3 decoy
-                    double decoyHits = 0;
-                    double totalHits = 0;
-                    var hits = psm.BestMatchingBioPolymersWithSetMods.GroupBy(p => p.FullSequence);
-                    foreach (var hit in hits)
-                    {
-                        if (hit.First().IsDecoy)
-                        {
-                            decoyHits++;
-                        }
-                        totalHits++;
-                    }
-                    cumulativeDecoy += decoyHits / totalHits;
-                    cumulativeDecoyPerNotch[notch] += decoyHits / totalHits;
-                }
-                else
+                // Group all hypotheses by notch
+                ambigResultsGroupedByNotch.Clear();
+                ambigResultsGroupedByNotch.AddRange(psm.BestMatchingBioPolymersWithSetMods
+                    .GroupBy(h => h.Notch));
+
+                // Aggregate FDR Counts
+                double totalHits = psm.BestMatchingBioPolymersWithSetMods.Count();
+                double totalTargetHits = 0;
+                double totalDecoyHits = 0;
+
+                foreach (var group in ambigResultsGroupedByNotch)
                 {
-                    cumulativeTarget++;
-                    cumulativeTargetPerNotch[notch]++;
+                    int notch = group.Key;
+                    double groupTargetHits = group.Count(h => !h.IsDecoy);
+                    double groupDecoyHits = group.Count(h => h.IsDecoy);
+
+                    totalTargetHits += groupTargetHits;
+                    totalDecoyHits += groupDecoyHits;
+
+                    cumulativeTargetPerNotch[notch] += groupTargetHits / totalHits;
+                    cumulativeDecoyPerNotch[notch] += groupDecoyHits / totalHits;
```

Comment:
Should the denominator here be groupHits?

Reply:
No. Total hits refers to all hits for this given PSM. For each notch, we are adding:
 
cumulativeTargetNotch += Targets in Notch / Ambiguous Result Count for PSM
cumulativeDecoyNotch += Decoys in Notch / Ambiguous Result Count for PSM